Bug 1304375 - no error in UI when release cannot be deleted

tieu opened the pull request:
The error message when deleting a release with rules pointing to it now shows a more comprehensive error. The response is now a 400 status when the error occurs.

Overview of the changes:

- Added new error response "HasRulesError"
- Modified UI Release Delete controller to display exceptions instead of blank message when errors occur.

mozbhearsum commented on auslib/test/admin/views/test_releases.py:
```diff
@@ -341,13 +341,6 @@ def testReleasePostInvalidBlob(self):
-        data = json.dumps(dict(bouncerProducts=dict(partial='foo'), name='a', hashFunction="sha512"))
-        ret = self._post("/releases/a", data=dict(data=data, product="a", data_version=1, schema_version=1))
-        self.assertStatusCode(ret, 400)
-        self.assertIn("This change requires signoff", ret.data)
```
Is there a reason you removed the this test, and the assert from the previous one?

tieu commented:
Thanks for the review, I will make the suggested changes now.

Overview of the new changes:

Removed error response "HasRulesError"
Added a new test case

mozbhearsum commented on auslib/admin/views/base.py:
```diff
@@ -51,6 +51,10 @@ def decorated(*args, **kwargs):
+            except ValueError as e:
+                msg = "Error. {}".format(e.message)
+                logging.warning(msg)
+                return Response(status=403, response=json.dumps({"exception": msg}), mimetype="application/json")
```
This should be a 400 actually - 403 is specifically for cases where somebody is not authorized to do something (eg, because of permissions).

mozbhearsum commented on auslib/test/admin/views/test_releases.py:
```diff
@@ -437,6 +437,10 @@ def testDeleteReadOnlyRelease(self):
         self.assertStatusCode(ret, 403)
+        
+    def testDeleteWithRules(self):
+        ret = self._delete("/releases/d", qs=dict(name='ab', data_version=1))
```
Two things here:
1) No need to include "name" in the query string - it's part of the URL.
2) As far as I can tell, "d" is not mapped to. This test is failing because it returns 200! You'll need to update this to choose one of the releases mentioned by the rules from https://github.com/mozilla/balrog/blob/master/auslib/test/admin/views/base.py#L82

mozbhearsum commented:
Hi @tieu, are you still looking at this?

aksareen commented:
Hey  @tieu, great job. 
Your checks are failing due to " auslib/test/admin/views/test_releases.py:440:1: W293 blank line contains whitespace ".
For future reference : try running "tox" or "./run-tests.sh backend" cli commands to see if the all PEP8 style checks and the tests succeed every time before pushing the changes.

aksareen commented on ui/app/js/controllers/release_delete_controller.js:
```diff
@@ -17,7 +17,7 @@ function ($scope, $modalInstance, CSRF, Releases, release, releases) {
       .error(function(response) {
          if (typeof response === 'object') {
-          message = '';
+          message = response.exception;
```
Hey @tieu, why are you setting default value to exception ?

aksareen replied:
As discussed on IRC with tieu, the exception message in response is being assigned to the variable to be displayed on UI. Previously the message was set as blank, hence the UI wasn't displaying the error message.

mozbhearsum replied:
Nice catch!
